### ISSUE TITLE
LearnerBuilder is too difficult to use

### ISSUE THREAD

**Issue by OstrichRunning** [tracel-ai/burn#3379]
```
match build_optimizer(self.config.optimizer(), &self.model) {
            AnyOptimizer::Sgd(optim) => {
                if self.config.metric().is_some() {
                    self.train_with_sdg_config(train_loader, val_loader, scheduler, device, optim);
                } else {
                    self.train_with_sdg(train_loader, val_loader, scheduler, device, optim);
                }
            }
            AnyOptimizer::Adam(optim) => {
                if self.config.metric().is_some() {
                    self.train_with_adam_config(train_loader, val_loader, scheduler, device, optim);
                } else {
                    self.train_with_adam(train_loader, val_loader, scheduler, device, optim);
                }
            }
            AnyOptimizer::AdamW(optim) => {
                if self.config.metric().is_some() {
                    self.train_with_adamw_config(train_loader, val_loader, scheduler, device, optim);
                } else {
                    self.train_with_adamw(train_loader, val_loader, scheduler, device, optim);
                }
            }
            AnyOptimizer::AdaGrad(optim) => {
                if self.config.metric().is_some() {
                    self.train_with_adagrad_config(train_loader, val_loader, scheduler, device, optim);
                } else {
                    self.train_with_adagrad(train_loader, val_loader, scheduler, device, optim);
                }
            }
        }
```
The above is my code. I hope to configure it more configurably and flexibly adapt to Optimizers such as sdg adam. Due to the LearnerBuilder type problem, do I need to repeatedly implement multiple LearnerBuilder::new(self.directory.clone())?

**Comment by OstrichRunning** [comment]
LearnerBuilder::new(self.directory.clone()) error：
the following types implement trait `ItemLazy`:
  ClassificationOutput<B>
  MultiLabelClassificationOutput<B>
  RegressionOutput<B>
  burn::train::metric::processor::base::LearnerItem<T> 

Various errors, too difficult to adapt, the delay is much longer than the custom train

**Comment by OstrichRunning** [comment]
<img width="1293" height="586" alt="Image" src="https://github.com/user-attachments/assets/8196bb1b-74c8-4c0d-a9ab-6b73c9ff4d65" />

**Comment by OstrichRunning** [comment]
I used version 0.17.1 and got many errors. I downloaded the demo and saw that the version was 0.18.0 and there were no errors. Was it fixed? But no new version was released?

**Comment by OstrichRunning** [comment]
<img width="2366" height="371" alt="Image" src="https://github.com/user-attachments/assets/e41352a8-d630-40cb-85ab-ee1320360a4f" />
The same code, the demo on the left and my 0.17.1 version on the right

**Comment by laggui** [comment]
> I used version 0.17.1 and got many errors. I downloaded the demo and saw that the version was 0.18.0 and there were no errors. Was it fixed? But no new version was released?

After a we release a new version we bump version used in the repository. We just release 0.18.0 though 🙂 (and the repository now points to 0.19.0)

It's a bit difficult to decipher what your actual compilation error is with the screenshots and all, but I suspect it's simply because the compiler cannot infer what the model output type is. 

The compiler should be able to automatically infer the type if you provide the model to the builder and then call `learner.fit(...)`. The [TrainStep trait bound is tied to the model passed to the learner at this point](https://github.com/tracel-ai/burn/blob/main/crates/burn-train/src/learner/train_val.rs#L122). So it should correctly see that the output type is `ClassificationOutput<B>` (well, actually I'm not sure what your output type is in this case). The adaptor is automatically implemented for `ClassificationOutput<B>` (as you can see in the compiler error message), but not sure what output type you are using.
Otherwise you have to specify what those types are, e.g.,

```rust
LearnerBuilder::<B, ClassificationOutput<B>, ClassificationOutput<B::InnerBackend>, _, _, _,>::new(...)
```

Thankfully we basically never have to write this out.. would be quite verbose 😅 the compiler is able to infer them. But when it's ambiguous, you get a compiler error stating that the concrete type needs to be specified.

**Comment by OstrichRunning** [comment]
I downloaded the demo and implemented my function according to modern-lstm. It has local dependencies like the demo. So far, no errors have been reported.
